### PR TITLE
Improve validate challenge config API

### PR DESCRIPTION
**Pull request by gchhablani**
This PR adds checks for ID validation in leaderboard, challenge phases and dataset splits.

### CODE REVIEW

File: apps/challenges/challenge_config_utils.py [hunk 1]
```diff
@@ -189,15 +201,19 @@ def is_challenge_phase_split_mapping_valid(
...
-                return True
-            else:
-                return False
-        else:
-            return False
-    return False
+    error_messages = []
+
+    if leaderboard_id not in leaderboard_ids:
+        error_messages.append("ERROR: Challenge phase split {} found leaderboard id {}. This relationship is not valid. Please check your configuration.".format(challenge_phase_split_index, leaderboard_id))
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:207]: Change message to `ERROR: Invalid leaderboard id {} found in challenge phase spit {}`

File: apps/challenges/challenge_config_utils.py [hunk 2]
```diff
@@ -189,15 +201,19 @@ def is_challenge_phase_split_mapping_valid(
...
+    if phase_id not in phase_ids:
+        error_messages.append("ERROR: Challenge phase split {} found phase id {}. This relationship is not valid. Please check your configuration.".format(challenge_phase_split_index, phase_id))
+    if dataset_split_id not in dataset_split_ids:
+        error_messages.append("ERROR: Challenge phase split {} found dataset split id {}. This relationship is not valid. Please check your configuration.".format(challenge_phase_split_index, dataset_split_id))
+
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:212]: rephrase error message similar to above

File: apps/challenges/challenge_config_utils.py [hunk 3]
```diff
@@ -353,6 +370,7 @@ def validate_challenge_config_util(
         error_messages.append(message)
 
     if not len(error_messages):
+        # TODO: Understand what is happening here
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:373]: remove

File: apps/challenges/challenge_config_utils.py [hunk 4]
```diff
@@ -370,6 +388,23 @@ def validate_challenge_config_util(
...
+    if existing_challenge:
+        existing_challenge_phases = ChallengePhase.objects.filter(challenge=existing_challenge.id)
+        existing_challenge_phase_splits = ChallengePhaseSplit.objects.filter(challenge_phase__in=existing_challenge_phases)
+        existing_leaderboards = Leaderboard.objects.filter(id__in=existing_challenge_phase_splits.values('leaderboard'))
+        existing_dataset_splits = DatasetSplit.objects.filter(id__in=existing_challenge_phase_splits.values('dataset_split'))
+
+        existing_leaderboard_config_ids = [int(x.config_id) for x in existing_leaderboards]
+        existing_dataset_split_config_ids = [int(x.config_id) for x in existing_dataset_splits]
+        existing_challenge_phase_config_ids = [int(x.config_id) for x in existing_challenge_phases]
+        existing_challenge_phase_split_id_triples = [(split.leaderboard.config_id, split.challenge_phase.config_id, split.dataset_split.config_id) for split in existing_challenge_phase_splits]
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:401]: rename variable and make code readable

File: apps/challenges/challenge_config_utils.py [hunk 5]
```diff
@@ -387,7 +422,10 @@ def validate_challenge_config_util(
             message = "ERROR: There is no 'labels' key in leaderboard schema."
             error_messages.append(message)
             error = True
-
+            # TODO: What to do if the labels is an empty list?
+            # TODO: Do we need to check for metric metadata?
+            # TODO: How to check for change when we add a leaderboard with new/same id - currently no error.
+            # TODO: Can check for ID repeats at least
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:428]: remove comment

File: apps/challenges/challenge_config_utils.py [hunk 6]
```diff
@@ -400,11 +438,19 @@ def validate_challenge_config_util(
                     )
                     error_messages.append(message)
                 else:
+                    # Addition
+                    if existing_leaderboard_config_ids and int(data["id"]) not in existing_leaderboard_config_ids:
+                        error_messages.append("ERROR: The leaderboard {} was not present in the existing leaderboard IDs. It is not possible to add new leaderboards once challenge is created. Please only use existing leaderboard IDs.".format(data["id"]))
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:443]: is flake8 not running on this? This single line is too long it seems.

Change message to `Leaderboard {} doesn't exist. Addition of new leaderboard after challenge creation is not allowed`

File: apps/challenges/challenge_config_utils.py [hunk 7]
```diff
@@ -400,11 +438,19 @@ def validate_challenge_config_util(
...
                     leaderboard_ids.append(data["id"])
     else:
         message = "ERROR: There is no key leaderboard in the YAML file."
         error_messages.append(message)
 
+    # Deletion Case:
+    for existing_leaderboard_id in existing_leaderboard_config_ids:
+        if existing_leaderboard_id not in leaderboard_ids:
+            error_messages.append("ERROR: The leaderboard {} was present in the existing leaderboard IDs but not found in the current configuration. It is not possible to delete leaderboards once challenge is created. Please use all the existing leaderboard IDs.".format(existing_leaderboard_id))
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:452]: rephrase error message similar to above

File: apps/challenges/challenge_config_utils.py [hunk 8]
```diff
@@ -413,8 +459,13 @@ def validate_challenge_config_util(
         return error_messages, yaml_file_data, files
 
     phase_ids = []
+    phase_codenames = []
     files["challenge_test_annotation_files"] = []
     for data in challenge_phases_data:
+        if data["codename"] not in phase_codenames:
+            phase_codenames.append(data["codename"])
+        else:
+            error_messages.append("Found codename `{}` in phase `{}` which in not unique. Please use unique codenames for all phases.".format(data["codename"], data["name"]))
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:468]: change to `ERROR: Duplicate codename {} for phase {}. Please ensure codenames are unique`

File: apps/challenges/challenge_config_utils.py [hunk 9]
```diff
@@ -529,10 +580,19 @@ def validate_challenge_config_util(
             )
             error_messages.append(message)
         else:
+            # Addition
+            if existing_challenge_phase_config_ids and int(data["id"]) not in existing_challenge_phase_config_ids:
+                error_messages.append("ERROR: The challenge phase {} was not present in the existing challenge phase IDs. It is not possible to add new challenge phases once challenge is created. Please only use existing challenge phase IDs.".format(data["id"]))
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:585]: rephrase error message similar to previous suggestions

File: apps/challenges/challenge_config_utils.py [hunk 10]
```diff
@@ -529,10 +580,19 @@ def validate_challenge_config_util(
...
             phase_ids.append(data["id"])
 
+    # Deletion Case:
+    for existing_challenge_phase_id in existing_challenge_phase_config_ids:
+        if existing_challenge_phase_id not in phase_ids:
+            error_messages.append("ERROR: The challenge phase {} was present in the existing challenge phase IDs but not found in the current configuration. It is not possible to delete challenge phases once challenge is created. Please use all the existing challenge phases IDs.".format(existing_challenge_phase_id))
+
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:592]: same here

File: apps/challenges/challenge_config_utils.py [hunk 11]
```diff
@@ -544,7 +604,10 @@ def validate_challenge_config_util(
                     )
                 )
                 error_messages.append(message)
-
+            if split["codename"] not in dataset_split_codenames:
+                dataset_split_codenames.append(data["codename"])
+            else:
+                error_messages.append("Found codename `{}` in dataset split `{}` which in not unique. Please use unique codenames for all dataset splits.".format(split["codename"], split["name"]))
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:610]: same here

File: apps/challenges/challenge_config_utils.py [hunk 12]
```diff
@@ -556,26 +619,35 @@ def validate_challenge_config_util(
                 )
                 error_messages.append(message)
             else:
+                # Addition
+                if existing_dataset_split_config_ids and int(split["id"]) not in existing_dataset_split_config_ids:
+                    error_messages.append("ERROR: The dataset split ID {} was not present in the existing dataset split IDs. It is not possible to add new dataset split once challenge is created. Please only use existing data split IDs.".format(split["id"]))
                 dataset_splits_ids.append(split["id"])
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:625]: same here

File: apps/challenges/challenge_config_utils.py [hunk 13]
```diff
@@ -556,26 +619,35 @@ def validate_challenge_config_util(
...
-
     else:
         message = "ERROR: There is no key for dataset splits."
         error_messages.append(message)
 
+    # Deletion Case:
+    for existing_dataset_split_config_id in existing_dataset_split_config_ids:
+        if existing_dataset_split_config_id not in dataset_splits_ids:
+            error_messages.append("ERROR: The dataset split {} was present in the existing dataset split IDs but not found in the current configuration. It is not possible to delete dataset splits once challenge is created. Please use all the existing dataset split IDs.".format(existing_dataset_split_config_id))
+
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:634]: same here

File: apps/challenges/challenge_config_utils.py [hunk 14]
```diff
@@ -556,26 +619,35 @@ def validate_challenge_config_util(
...
-            ):
-                message = (
-                    "ERROR: Challenge phase split {} has invalid keys "
-                    "for challenge_phase_id, leaderboard_id, dataset_split_id"
-                ).format(phase_split)
-                error_messages.append(message)
+            if (data["leaderboard_id"], data["challenge_phase_id"], data["dataset_split_id"]) not in existing_challenge_phase_split_id_triples:
+                error_messages.append("ERROR: You are trying to create a new challenge phase split with (leaderboard_id: {}, challenge_phase_id: {}, dataset_split_id: {}) which is not present in existing phase splits. Adding phase splits is not allowed.".format(data["leaderboard_id"], data["challenge_phase_id"], data["dataset_split_id"]))
+            else:
+                challenge_phase_split_id_triples.append((data["leaderboard_id"], data["challenge_phase_id"], data["dataset_split_id"]))
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:645]: same here

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: apps/challenges/challenge_config_utils.py (hunk 4 above)
Comment (gchhablani) [reply apps/challenges/challenge_config_utils.py]: I have updated it but if you have any suggestions please let me know.

File: apps/challenges/challenge_config_utils.py [hunk 15]
```diff
@@ -209,7 +221,7 @@ def get_value_from_field(data, base_location, field_name):
 
 
 def validate_challenge_config_util(
-    request, challenge_host_team, BASE_LOCATION, unique_folder_name, zip_ref
+    request, challenge_host_team, BASE_LOCATION, unique_folder_name, zip_ref, e_challenge
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:224]: Can you rename the variable `e_challenge` to `current_challenge`

File: apps/challenges/challenge_config_utils.py [hunk 16]
```diff
@@ -370,6 +383,23 @@ def validate_challenge_config_util(
...
+
+        exis_leaderboard_config_ids = [int(x.config_id) for x in exis_leaderboards]
+        exis_dataset_config_ids = [int(x.config_id) for x in exis_dataset_splits]
+        exis_phase_config_ids = [int(x.config_id) for x in exis_challenge_phases]
+        exis_phase_split_ids = [(split.leaderboard.config_id, split.challenge_phase.config_id, split.dataset_split.config_id) for split in exis_challenge_phase_splits]
+    else:
+        exis_leaderboard_config_ids = []
+        exis_dataset_config_ids = []
+        exis_phase_config_ids = []
+        exis_phase_split_ids = []
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:401]: same here, use the `current_<var_name` for all of these

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: apps/challenges/challenge_config_utils.py [hunk 17]
```diff
@@ -400,11 +429,19 @@ def validate_challenge_config_util(
                     )
                     error_messages.append(message)
                 else:
+                    # Addition
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:432]: remove comment

File: apps/challenges/challenge_config_utils.py [hunk 18]
```diff
@@ -400,11 +429,19 @@ def validate_challenge_config_util(
...
+                    if current_leaderboard_config_ids and int(data["id"]) not in current_leaderboard_config_ids:
+                        error_messages.append("ERROR: Leaderboard {} doesn't exist. Addition of new leaderboard after challenge creation is not allowed.".format(data["id"]))
                     leaderboard_ids.append(data["id"])
     else:
         message = "ERROR: There is no key leaderboard in the YAML file."
         error_messages.append(message)
 
+    # Deletion Case:
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:440]: remove comment. The error message explains the purpose of the check

File: apps/challenges/challenge_config_utils.py [hunk 19]
```diff
@@ -529,10 +571,19 @@ def validate_challenge_config_util(
             )
             error_messages.append(message)
         else:
+            # Addition
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:574]: remove

File: apps/challenges/challenge_config_utils.py [hunk 20]
```diff
@@ -529,10 +571,19 @@ def validate_challenge_config_util(
...
+            if current_phase_config_ids and int(data["id"]) not in current_phase_config_ids:
+                error_messages.append("ERROR: Challenge phase {} doesn't exist. Addition of new challenge phase after challenge creation is not allowed.".format(data["id"]))
             phase_ids.append(data["id"])
 
+    # Deletion Case:
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:579]: remove

File: apps/challenges/challenge_config_utils.py [hunk 21]
```diff
@@ -556,26 +610,35 @@ def validate_challenge_config_util(
                 )
                 error_messages.append(message)
             else:
+                # Addition
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:613]: remove

File: apps/challenges/challenge_config_utils.py [hunk 22]
```diff
@@ -556,26 +610,35 @@ def validate_challenge_config_util(
...
+                if current_dataset_config_ids and int(split["id"]) not in current_dataset_config_ids:
+                    error_messages.append("ERROR: Dataset split {} doesn't exist. Addition of new dataset split after challenge creation is not allowed.".format(split["id"]))
                 dataset_splits_ids.append(split["id"])
-
     else:
         message = "ERROR: There is no key for dataset splits."
         error_messages.append(message)
 
+    # Deletion Case:
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:621]: remove

File: apps/challenges/challenge_config_utils.py [hunk 23]
```diff
@@ -556,26 +610,35 @@ def validate_challenge_config_util(
...
+    for current_dataset_split_config_id in current_dataset_config_ids:
+        if current_dataset_split_config_id not in dataset_splits_ids:
+            error_messages.append("ERROR: Dataset split {} not found in config. Deletion of existing dataset split after challenge creation is not allowed.".format(current_dataset_split_config_id))
+
     # Check for challenge phase splits
     challenge_phase_splits = yaml_file_data.get("challenge_phase_splits")
+    challenge_phase_split_id_triples = []
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:628]: rename to `challenge_phase_split_uuids`

File: apps/challenges/challenge_config_utils.py [hunk 24]
```diff
@@ -587,6 +650,10 @@ def validate_challenge_config_util(
                 )
                 error_messages.append(message)
             phase_split += 1
+
+        for triple in current_phase_split_ids:
```
Comment (Ram81) [inline apps/challenges/challenge_config_utils.py:654]: `for uuid in current_phase_split_ids`

**Review by Ram81 — CHANGES_REQUESTED** [review]
minor changes

**Pull request merged**